### ISSUE TITLE
Generated with function has incorrect type

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#745]
```agda
module With where

-- This is all standard library stuff, inspect on steroids:

data _≡_ {A : Set} (x : A) : A → Set where
  refl : x ≡ x

data Unit : Set where
  unit : Unit

Hidden : Set → Set
Hidden A = Unit → A

hide : {A : Set} {B : A → Set} →
       ((x : A) → B x) → ((x : A) → Hidden (B x))
hide f x unit = f x

reveal : {A : Set} → Hidden A → A
reveal f = f unit

data Reveal_is_ {A : Set} (x : Hidden A) (y : A) : Set where
  [_] : (eq : reveal x ≡ y) → Reveal x is y

inspect : {A : Set} {B : A → Set}
          (f : (x : A) → B x) (x : A) → Reveal (hide f x) is (f x)
inspect f x = [ refl ]

data ℕ : Set where
  zero : ℕ
  suc : ℕ -> ℕ


-- New stuff starts here:

data Wrap : ℕ -> Set where
  con : (n : ℕ) -> Wrap n

wrap : (n : ℕ) -> Wrap n
wrap zero    = con zero
wrap (suc x) = con (suc x)

bar-with : (n : ℕ)(v : Wrap n) -> Reveal (hide wrap n) is (wrap n) -> ℕ
bar-with n (con .n) r = zero

bar : (n : ℕ) -> ℕ
bar zero = zero
bar (suc n) = bar-with n (wrap n) (inspect wrap n)

-- I've manually desugared `bar' to make the error clearer, but the
-- following definition works correctly:
{-
bar : (n : ℕ) -> ℕ
bar zero = zero
bar (suc n) with wrap n | inspect wrap n
bar (suc ._) | con _    | r = zero
-}

foo : (n : ℕ) -> Wrap (bar n) -> ℕ
foo zero p = zero
foo (suc x) p with inspect bar x
foo (suc x) p | r = zero

-- The previous line gives the error:

-- ℕ != Wrap x of type Set
-- when checking that the type
-- (x : ℕ) (w : Reveal_is_ {ℕ} (hide {ℕ} {λ _ → ℕ} bar x) (bar x))
-- (p : Wrap (bar-with x (wrap x) w)) →
-- ℕ
-- of the generated with function is well-formed

-- In particular, notice that the type of `p' is incorrect, because an
-- occurrence of `inspect wrap x` has been turned into the variable w
-- (which corresponds to `inspect bar x`).

-- I think the correct desugaring of the with is as follows:

good-foo-with : (x : ℕ) -> Wrap (bar (suc x)) -> Reveal (hide bar x) is bar x 
-> ℕ
good-foo-with x w r = zero

good-foo : (n : ℕ) -> Wrap (bar n) -> ℕ
good-foo zero p = zero
good-foo (suc x) p = good-foo-with x p (inspect bar x)
```

Original issue reported on code.google.com by `adamgundry` on 2 Nov 2012 at 10:33

**Comment by UlfNorell** [comment]
Fixed by reconstructing constructor parameters before with-abstraction.

@WolframKahl: Can you check this on your code. There's a small but noticeable performance penalty for this fix on the standard library. Might be a lot worse for you.

Important for #1845.

**Comment by WolframKahl** [comment]
A lot of my with-heavy code typechecked with 4ca0087feeee1e6d0bcba4a46e85f39c988f47ff, in 202 seconds, compared with 182 seconds for  4ca0087feeee1e6d0bcba4a46e85f39c988f47ff, but for being able to check everything I am blocked on #1726.

More detailed timing results, including `-v profile:7`, in the attached file.

[2016-02-22_Agda-2.4.3.22_ListSetMap_log.txt](https://github.com/agda/agda/files/141910/2016-02-22_Agda-2.4.3.22_ListSetMap_log.txt)

Wolfram

**Comment by UlfNorell** [comment]
Thanks for testing Wolfram. From the looks of things I think this fix is too expensive. I'll see about fixing it in a less principled, but more efficient way.

**Comment by UlfNorell** [comment]
I'm coming up empty on alternative fixes. A work-around might be to make one of the parameters to the inspect type an index.

**Comment by WolframKahl** [comment]
In my eyes, that slow-down as such is harmless ─ recent Agda releases were more likely to have speedups in at least that order of magnitude, and doing the right thing has priority.
(I would probably be more worried if it used 10% more heap... :wink: ).

Once my other code typechecks again, I will post numbers for that, too (assuming you re-introduce that fix) ─ AContext, for example, does not contain a single `with`.

**Comment by UlfNorell** [comment]
I have an idea for a more efficient, but still correct fix. Stay tuned...

**Comment by UlfNorell** [comment]
The fix in d93cba2 should be both correct and have very little performance impact.

**Comment by WolframKahl** [comment]
Something changed: At 84a05643b97140cbe7a6728155b13c1a499efb34, with standard library as sub-module at dcc2ab74e066dc15cfcfcee3e93a474939ecba95, I get

```
time agda -v profile:7 +RTS -S -K256M -H11G -M11G -A1G -RTS -i . -i $(agdalib) Data/SUList/mapMImage.lagda > mapMImage_log.txt 2>&1
[...]
/var/tmp/kahl/svn/A/RATH-Agda/trunk/mapMImage/Data/SUList/mapMImage.lagda:111,7-116,114
f a bs != _ of type Maybe SetC.ListSet₁
when checking that the clause
from-∈-mapMImage₀ {f} {_} {_∶≈_∶<_<∶_ (a , bs) {k} k≈a {m} k<m es}
  a₀
  c₀
  (a₀∈MR , c₀∈a₀MR)
  with f a bs | inspect (f a) bs
from-∈-mapMImage₀ {f} {_} {_∶≈_∶<_<∶_ (a , bs) {k} k≈a {m} k<m es}
  a₀
  c₀
  (a₀∈MR , c₀∈a₀MR)
  | nothing | _
  = Product.map₁ (Map.∈tail _ _ _)
    (from-∈-mapMImage₀ {f} {_} {es} a₀ c₀ (a₀∈MR , c₀∈a₀MR))
from-∈-mapMImage₀ {f} {_} {_∶≈_∶<_<∶_ (a , bs) {k} k≈a {m} k<m es}
  a₀
  c₀
  (Map.∈head _ _ a₀≡k _ _ , c₀∈a₀MR)
  | just cs | ≡[ eq ]≡
  = Map.∈head _ _ a₀≡k _ _ , cs , c₀∈a₀MR , eq
from-∈-mapMImage₀ {f} {_} {_∶≈_∶<_<∶_ (a , bs) {k} k≈a {m} k<m es}
  a₀
  c₀
  (Map.∈tail _ _ _ a₀∈es , c₀∈a₀MR)
  | just cs | _
  = Product.map₁ (Map.∈tail _ _ _)
    (from-∈-mapMImage₀ {f} {_} {es} a₀ c₀ (a₀∈es , c₀∈a₀MR))
has type
{f : A₀ → SetB.ListSet₁ → Maybe SetC.ListSet₁} {m : Maybe A₀}
{R
 : .Data.SUList.RawListSetMap.Core.RawListSetMapCore.Map.SUList A B
   m}
(a₀ : A₀) (c₀ : C₀) →
(A .Data.SUList.RawListSetMap.Core.RawListSetMapCore.∈ C) (a₀ , c₀)
(mapMImage f (m , R)) →
Σ∶• ((A Map.∈ B) a₀ R)
(λ a₀∈R →
   Σ∶• SetC.ListSet₁
   (λ cs → c₀ SetC.∈ cs × uncurry f (Map.∈-Elem′ A B a₀∈R) ≡ just cs))
```

for something that still worked with b0aaf8841badf5b843acba56582d6d53cb5ad083.

I couldn't even figure out which underscore this error message is referring to, so I just created a slice of the two top-level modules and packaged all dependencies.

[mapMImage_log.txt](https://github.com/agda/agda/files/145870/mapMImage_log.txt)

[mapMImage.tar.gz](https://github.com/agda/agda/files/145871/mapMImage.tar.gz)

**Comment by UlfNorell** [comment]
I changed with-abstraction to be type-aware to fix the problem of abstracting over terms of the wrong type. The problem with that is that you have to be careful to keep all the types and contexts lined up. In this case we're abstracting over `inspect (f a) bs` after having first abstracted over `f a bs`, but we do this in the original context instead of the context abstracted over `f a bs`. I'll see what I can do...

**Comment by UlfNorell** [comment]
Here's a cut-down example:

``` agda
postulate
  U V : Set
  T   : V → Set
  H   : ∀ v → T v → Set
  f   : U → V

data D : Set where
  d : D

foo : (u : U) (t : T (f u)) (h : H (f u) t) → Set
foo u t h with f u | d
... | _ | _ = ?
```

It turned out that the aggressive with introduced in d8b1317 was playing fast and loose with the well-formedness of intermediate types and telescopes, so I had to revert that to make this work. The good thing about that is that it vastly simplified the with-machinery.

**Comment by WolframKahl** [comment]
Just a quick heads-up for the speed: What I call Agda-2.4.3.24 (4a849c4404ef78a57bfebf75516b82e8ed8df59d) was actually faster on my with-heavy code than  4ca0087feeee1e6d0bcba4a46e85f39c988f47ff (even though it allocates a bit more and generates slightly larger .agdai files).

(These times are on a slower machine than those posted previously in this issue.)

```
  Alloc (MB)  Max.Res.(MB)  AllocRate   real    productivity  agdai
Data/SUList/ListSetMap.lagda  from scratch (71 modules, lots of ``with''):
3.20:   309,432    2,207   1,042,155,849   5m52.020s  84.6%  31,334,213
3.21:   307,994    2,263   1,029,974,316   5m52.669s  84.9%  31,334,013
3.24:   319,372    2,366   1,099,989,330   5m44.149s  84.5%  31,429,046
```

Agda is busy typechecking more of my stuff...

**Comment by WolframKahl** [comment]
I have not encountered any trouble with 4a849c4404ef78a57bfebf75516b82e8ed8df59d (2016-02-25); speed-wise there are also no big changes compared with 4ca0087feeee1e6d0bcba4a46e85f39c988f47ff (2016-02-06). The `-v profile:7` output of the following runs is contained in the attached file.

[2016-02-25_Agda-2.4.3.24.log.txt](https://github.com/agda/agda/files/149695/2016-02-25_Agda-2.4.3.24.log.txt)

```
FIRST SUMMARY LINES:
Agda-2.4.3.24 is 4a849c4404ef78a57bfebf75516b82e8ed8df59d pulled 2016-02-25,
compiled with ghc-7.10.3; with standard library pulled as sub-module:
dcc2ab74e066dc15cfcfcee3e93a474939ecba95

SECOND SUMMARY LINES:
Agda-2.4.3.20 is 4ca0087feeee1e6d0bcba4a46e85f39c988f47ff pulled 2016-02-06,
compiled with ghc-7.10.3.

THIRD SUMMARY LINES:
Agda-2.4.3.21 is 4ca0087feeee1e6d0bcba4a46e85f39c988f47ff pulled 2016-02-06
with Andrea Vezzosi's patch from
  https://github.com/agda/agda/issues/1625#issuecomment-132196576
, compiled with ghc-7.10.3.

Run on schroeder (Phenom II 3.20GHz 16GB) without --sharing.

Summary:
========

  Alloc (MB)  Max.Res.(MB)  AllocRate   real    productivity  agdai
Data/SUList/ListSetMap.lagda  from scratch (71 modules, lots of ``with''):
3.24:   319,372   2,366   1,099,989,330   5m44.149s  84.5%  31,429,046
3.20:   309,432   2,207   1,042,155,849   5m52.020s  84.6%  31,334,213
3.21:   307,994   2,263   1,029,974,316   5m52.669s  84.9%  31,334,013

Data/SUList/ListSetMap/FinRel.lagda  from scratch:
3.24    998,375   5,377   1,046,847,946  19m20.582s  82.5%  91,503,442
3.20  1,041,696   7,000   1,016,065,863  21m34.089s  79.6%  92,387,872  
3.21    982,497   6,233   1,000,457,455  20m44.568s  79.2%  92,393,715

Data.ADHG3.GSMonCat after FinRel:
3.24  2,023,112   4,611   1,294,496,065  29m15.040s  89.1%  115,326,425
3.20  2,017,094   4,906   1,339,702,343  28m22.281s  88.5%  116,218,271
3.21    574,361   4,820   1,317,241,748   9m50.571s  74.5%  116,236,294

Main.lagda after FinRel (and Data.ADHG3.GSMonCat):
3.24    148,367   3,083     988,400,853   3m04.977s  81.4%  126,197,056
3.20    151,606   3,115     972,330,551   3m12.891s  81.1%  127,051,798
3.21     85,880   3,153     817,535,842   2m20.653s  75.0%  127,070,597

========================================================================
AContext2 (no with, and no zero-ary constructors) from scratch:
3.24  1,055,408   2,297   1,173,672,530  16m05.355s  93.2%   39,393,882
3.20  1,046,566   2,717   1,181,301,058  16m02.052s  92.1%   40,359,092
3.21    454,499   2,547   1,090,167,004   8m09.901s  85.2%   40,359,523

Categoric/OCC/CSL/FromToAContext2.lagda after AContext2:
3.24  9,301,271   1,437   1,217,883,035  << killed after 127 min. >> 99.5%
3.20 10,509,593   1,457   1,278,168,089  << killed after 137 min. >> 99.5%
3.21     18,433   1,408     779,617,707   0m31.685s  76.0%

========================================================================
Data.ADHG4.GSMonCat from scratch:
3.24   960,047    2,918   1,283,236,005  14m03.537s  88.7%   37,278,141
3.20   951,180    2,950   1,285,318,840  13m58.805s  88.3%   37,463,679
3.21   850,550    2,957   1,245,893,838  13m25.671s  84.8%   37,461,844

========================================================================
Main.lagda from scratch:
3.24 1,004,113    5,493   1,033,470,582  19m42.267s  82.5%   87,790,435
3.20 1,051,328    5,412   1,013,815,686  20m52.832s  83.1%   88,560,175
3.21   931,150    5,459     985,628,420  19m38.628s  80.5%   88,564,666

Compiling Main to Haskell --- MAlonzo only, no GHC:
3.24   240,148   3,466    1,208,531,788   4m34.999s  72.8%  59,436,517 .hs
3.20   239,468   3,485    1,284,496,194   4m23.872s  71.2%  59,513,986 .hs
3.21   239,334   3,492    1,267,942,163   4m25.859s  71.6%  59,397,963 .hs

Compiling Main.hs with GHC:
3.24   680,753   2,873      946,601,224  15m41.412s   76.5%  99,830,704 exec.
3.20   678,422   2,740      952,445,091  15m35.801s   76.2%  99,794,016 exec.
3.21   677,297   3,017      955,246,725  15m32.805s   76.3%  99,678,448 exec.

Running Main (best of about 10 runs):
3.24     1,490    0.117   1,081,299,906   0m01.461s   98.1%
3.20     1,490    0.117   1,130,905,450   0m01.397s   98.2%
3.21     1,490    0.117   1,129,693,818   0m01.399s   98.2%

==============================================================
```

**Comment by UlfNorell** [comment]
Excellent. Thanks for testing.

**Comment by andreasabel** [comment]
Fix introduced #2341.
